src/mc-fle2-rfds.c

```diff
@@ -0,0 +1,615 @@
+   }
+
+   enum { UNKNOWN, MATCH_EXPRESSION, AGGREGATE_EXPRESSION } arg_type = UNKNOWN;
+
+   while (bson_iter_next (&array)) {
+      bson_iter_t doc;
+
+      if (!BSON_ITER_HOLDS_DOCUMENT (&array)) {
+         ERR_WITH_BSON (in, "%s", "expected document in '$and' array");
+         goto fail;
+      }
+
+      if (!bson_iter_recurse (&array, &doc)) {
+         ERR_WITH_BSON (in, "%s", "failed to recurse into '$and' element");
+         goto fail;
+      }
+
+      if (!bson_iter_next (&doc)) {
+         ERR_WITH_BSON (in, "%s", "unexpected empty '$and' array document");
+         goto fail;
+      }
+
+      if (arg_type == UNKNOWN) {
+         // Attempt to determine argument type by inspecting first key.
+         if (is_supported_operator (bson_iter_key (&doc))) {
+            // Assume the document is part of an Aggregate Expression, like:
+            // {$gt: ["$age", 5]}
+            arg_type = AGGREGATE_EXPRESSION;
+            spec->isAggregateExpression = true;
+         } else {
+            // Assume the document is part of a Match Expression, like:
+            // {age: {$gt: 5}}
+            arg_type = MATCH_EXPRESSION;
+         }
+      }
+
+      operator_value_t op;
+      switch (arg_type) {
+      case AGGREGATE_EXPRESSION:
+         if (!parse_aggregate_expression (in, doc, &op, status)) {
+            goto fail;
+         }
+         break;
+      case MATCH_EXPRESSION:
+         if (!parse_match_expression (in, doc, &op, status)) {
+            goto fail;
+         }
+         break;
+      case UNKNOWN:
+      default:
+         ERR_WITH_BSON (in, "%s", "unexpected unknown expression type");
+         goto fail;
+      }
+
+      switch (op.op_type) {
+      case FLE2RangeOperator_kGt:
+      case FLE2RangeOperator_kGte:
+         if (spec->lower.set) {
+            ERR_WITH_BSON (in, "unexpected duplicate bound %s", op.op_type_str);
+            goto fail;
+         }
+         spec->lower.set = true;
+         spec->lower.value = op.value;
+         spec->lower.included = op.op_type == FLE2RangeOperator_kGte;
+         break;
+      case FLE2RangeOperator_kLt:
+      case FLE2RangeOperator_kLte:
+         if (spec->upper.set) {
+            ERR_WITH_BSON (in, "unexpected duplicate bound %s", op.op_type_str);
+            goto fail;
+         }
+         spec->upper.set = true;
+         spec->upper.value = op.value;
+         spec->upper.included = op.op_type == FLE2RangeOperator_kLte;
+         break;
+      case FLE2RangeOperator_kNone:
+      default:
+         ERR_WITH_BSON (in, "unsupported operator type %s", op.op_type_str);
+         goto fail;
+         break;
+      }
+
+      if (spec->field) {
+         if (0 != strcmp (spec->field, op.field)) {
+            ERR_WITH_BSON (
+               in,
+               "unexpected field mismatch. Expected all fields to be "
+               "%s, but got %s",
+               spec->field,
+               op.field);
+            goto fail;
+         }
+      }
```

Comment:
```suggestion
      if (spec->field && 0 != strcmp (spec->field, op.field)) {
         ERR_WITH_BSON (in,
                        "unexpected field mismatch. Expected all fields to be "
                        "%s, but got %s",
                        spec->field,
                        op.field);
         goto fail;
      }
```
Minor simplification of conditional expressions.